### PR TITLE
Add an aggregate metric for the theoretical write capacity

### PR DESCRIPTION
**Pull request by mkeeler**
The new metric emits a sample of the number of logs per second boltdb could store if all those log write operations looked like the one currently being measured. That means that another operation would have the same number of logs per batch and that the actual txn Commit took the same amount of time. While no two operations will be identical, taking the average of the sample/summary emitted should provide a good picture of what Consul could handle with the current types of write operations being performed.

It is expected that this value will fluctuate with changes in size of the logs flowing through consul and how many logs get batched into one storage op.

If someone wanted to monitor this I think they would want to know when the actual write rate exceeds 75% of this metrics value. That could be due to an increased number of writes, or a degradation in disk performance which causes similar writes to slow down. Regardless of the cause, if you are getting close to the limit or see a drastic change in the metric it could be indicative of another issue which requires investigation.

### CODE REVIEW

File: bolt_store.go [hunk 1]
```diff
@@ -194,6 +195,15 @@ func (b *BoltStore) StoreLogs(logs []*raft.Log) error {
...
 	metrics.AddSample([]string{"raft", "boltdb", "logsPerBatch"}, float32(len(logs)))
 	metrics.AddSample([]string{"raft", "boltdb", "logBatchSize"}, float32(batchSize))
+	// Both the deferral and the inline function are important for this metrics
+	// accuracy. Deferral allows us to calculate the metric after the tx.Commit
+	// has finished and thus account for all the processing of the operation.
+	// The inlined function ensures that we do not calculate the time.Since(now)
+	// at the time of deferral but rather when the go runtime executes the 
+	// deferred function.
+	defer func() {
+		metrics.AddSample([]string{"raft", "boltdb", "writeCapacity"}, (float32(1000000000) / float32(time.Since(now).Nanoseconds())) * float32(len(logs)))
```
Comment (markan) [inline bolt_store.go:205]: Maybe digit separators e.g. 1_000_000_000

File: bolt_store.go [hunk 2]
```diff
@@ -168,7 +168,8 @@ func (b *BoltStore) StoreLog(log *raft.Log) error {
 
 // StoreLogs is used to store a set of raft logs
 func (b *BoltStore) StoreLogs(logs []*raft.Log) error {
-	defer metrics.MeasureSince([]string{"raft", "boltdb", "storeLogs"}, time.Now())
+	now := time.Now()
+	defer metrics.MeasureSince([]string{"raft", "boltdb", "storeLogs"}, now)
```
Comment (markan) [inline bolt_store.go:172]: Any reason not to combine this and the later defer? It might make it a little easier to notice that the two calls are using the same time interval as their basis.

**Review by markan — APPROVED** [review]
Nice. I really appreciate it when software surfaces metrics that captures the 

A few minor tweaks, feel free to take or leave them.

File: bolt_store.go (hunk 1 above)
Comment (mkeeler) [reply bolt_store.go]: Done (also required bumping the version in go.mod). I took it to 1.16 as that is the oldest supported Go version.

File: bolt_store.go (hunk 2 above)
Comment (mkeeler) [reply bolt_store.go]: That is a good idea and I have moved it into the deferred func.

File: v2/bolt_store.go [hunk 3]
```diff
@@ -203,6 +204,16 @@ func (b *BoltStore) StoreLogs(logs []*raft.Log) error {
 
 	metrics.AddSample([]string{"raft", "boltdb", "logsPerBatch"}, float32(len(logs)))
 	metrics.AddSample([]string{"raft", "boltdb", "logBatchSize"}, float32(batchSize))
+	// Both the deferral and the inline function are important for this metrics
+	// accuracy. Deferral allows us to calculate the metric after the tx.Commit
+	// has finished and thus account for all the processing of the operation.
+	// The inlined function ensures that we do not calculate the time.Since(now)
+	// at the time of deferral but rather when the go runtime executes the
+	// deferred function.
```
Comment (FFMMM) [inline v2/bolt_store.go:212]: [nit/ learning for me] couldn't we  make this a func w some arguments like `time.now` from up top and the number of logs? and call it w `defer` ?

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,6 @@
 module github.com/hashicorp/raft-boltdb
 
-go 1.12
+go 1.16
```
Comment (FFMMM) [inline go.mod:3]: [nit]: could we make this a separate PR? unless, it's needed for this PR of course.

**Comment by FFMMM** [comment]
[nit] Any chance we could add some comments around what the metric `writeCapacity` is and how folks should interact with it? (the PR description is great imo for this purpose)

File: go.mod (hunk 4 above)
Comment (mkeeler) [reply go.mod:3]: It is needed to enable using `_` in literal integers.

File: v2/bolt_store.go (hunk 3 above)
Comment (mkeeler) [reply v2/bolt_store.go:212]: We certainly could. Originally I had done that but then instead thought it might be nice to not have to follow fn calls to get to metrics that are emitted.

There are some extremely minor advantages/disadvantages for both but I don't have strong feelings about which way is best.

**Comment by mkeeler** [comment]
I added some info in the README about how to interpret metrics.

File: v2/bolt_store.go (hunk 3 above)
Comment (mkeeler) [reply v2/bolt_store.go:212]: Given how many args we would have to pass through I don't think its worth it here.

File: bolt_store.go [hunk 5]
```diff
@@ -194,6 +194,16 @@ func (b *BoltStore) StoreLogs(logs []*raft.Log) error {
...
 	metrics.AddSample([]string{"raft", "boltdb", "logsPerBatch"}, float32(len(logs)))
 	metrics.AddSample([]string{"raft", "boltdb", "logBatchSize"}, float32(batchSize))
+	// Both the deferral and the inline function are important for this metrics
+	// accuracy. Deferral allows us to calculate the metric after the tx.Commit
+	// has finished and thus account for all the processing of the operation.
+	// The inlined function ensures that we do not calculate the time.Since(now)
+	// at the time of deferral but rather when the go runtime executes the
+	// deferred function.
+	defer func() {
+		metrics.AddSample([]string{"raft", "boltdb", "writeCapacity"}, (float32(1_000_000_000)/float32(time.Since(now).Nanoseconds()))*float32(len(logs)))
```
[the remark this answers is not in the document]
Comment (mkeeler) [reply bolt_store.go:204]: Not really, especially in any sort of CI where disk performance and such would be unpredictable.
